README.md
# JobSchedulers.jl

*A Julia-based job scheduler and workload manager inspired by Slurm and PBS.*

## Why JobScheduler?

We may find different tasks or programs use different CPU and memory. Some can run simultaneously, but some have to run sequentially. JobScheduler is stable, useful and powerful for task queuing and workload management. 

## Package Features

- Job and task scheduler.
- Local workload manager.
- Support CPU, memory, run time management.
- Support running a job at specific time, or a period after creating (schedule).
- Support deferring a job until specific jobs reach specific states (dependency).
- Support automatic backup and reload.
- Fancy progress meter in terminal.

  ![progress meter](assets/progress_meter.png)

## Future development

- Support command-line scheduler by using DaemonMode.jl.
- Use Documenter.jl for documentation.

## Installation

JobSchedulers.jl can be installed using the Julia package manager. From the Julia REPL, type ] to enter the Pkg REPL mode and run

```julia
pkg> add JobSchedulers
```

To use the package, type

```julia
using JobSchedulers
```

## Quick start

If you need to run multiple heavy Julia tasks, it is recommended to [start Julia with multi-threads](https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads).

```julia
using JobSchedulers, Dates
```

### Scheduler Controls

```julia
scheduler_start()
# [ Info: Scheduler starts.

scheduler_status()
# ┌ Info: Scheduler is running.
# │   SCHEDULER_MAX_CPU = 32
# │   SCHEDULER_MAX_MEM = 121278191616
# │   SCHEDULER_UPDATE_SECOND = 0.05
# │   JOB_QUEUE_MAX_LENGTH = 10000
# └   SCHEDULER_TASK = Task (runnable) @0x00007fe205052e60
# :running

# scheduler_stop()  # NO RUN
```

### Scheduler Settings

Set the **maximum CPU** that the scheduler can use. If starting Julia with multi-threads, the maximum CPU is `nthreads() - 1`.

```julia
set_scheduler_max_cpu()     # use all available CPUs
# 32
set_scheduler_max_cpu(4)    # use 4 CPUs
# 4
set_scheduler_max_cpu(0.5)  # use 50% of CPUs
# 16
```

Set the **maximum RAM** the scheduler can use:

```julia
set_scheduler_max_mem()             # use 80% of total memory
# 107792089088

set_scheduler_max_mem(4GB)          # use 4GB memory
set_scheduler_max_mem(4096MB)
set_scheduler_max_mem(4194304KB)
set_scheduler_max_mem(4294967296B)
# 4294967296
set_scheduler_max_mem(0.5)          # use 50% of total memory
# 67370055680
```

Set the update interval of job queue:

```julia
set_scheduler_update_second(0.03)  # update job queue every 0.3 seconds
# 0.03
```

Set the maximum number of finished jobs:

```julia
set_scheduler_max_job(10000)  # If number of finished jobs > 10000, the oldest ones will be removed.
# 10000                       # It does not affect queuing or running jobs.
```

Set the previous setting in one function:

```julia
set_scheduler(
    max_cpu = JobSchedulers.SCHEDULER_MAX_CPU,
    max_mem = JobSchedulers.SCHEDULER_MAX_MEM,
    update_second = JobSchedulers.SCHEDULER_UPDATE_SECOND,
    max_job = JobSchedulers.JOB_QUEUE_MAX_LENGTH
)
# ┌ Info: Scheduler is running.
# │   SCHEDULER_MAX_CPU = 16
# │   SCHEDULER_MAX_MEM = 67370055680
# │   SCHEDULER_UPDATE_SECOND = 0.03
# │   JOB_QUEUE_MAX_LENGTH = 10000
# └   SCHEDULER_TASK = Task (runnable) @0x00007fd239184fe0
# :running
```

### Job Controls

A `Job` is the wrapper of `AbstractCmd`, `Function` or `Task`:

```julia
command_job = Job(
    `echo command job done`    # AbstractCmd to run
)

function_job = Job() do  # the function should have no arguments
    println("function job done")
end

task_job = Job(
    @task(println("task job done"))  # Task to run
)

job_with_args = Job(
    @task(begin println("job_with_args done"); "result" end); # Task to run
    name = "job with args",               # job name.
    user = "me",                # Job owner.
    ncpu = 1,                   # Number of CPU required.
    mem = 1KB,                  # Number of memory required (unit: TB, GB, MB, KB, B).
    schedule_time = Second(3),  # Run after 3 seconds; can be ::DateTime or ::Period.
    wall_time = Hour(1),        # The maximum time to run the job. (Cancel job after reaching wall time.)
    priority = 20,              # Lower number = higher priority.
    dependency = [              # Defer job until some jobs reach some states.
        DONE => command_job,
        DONE => task_job.id
    ]
)
# Job:
#   id            → 3603962603799412
#   name          → "job with args"
#   user          → "me"
#   ncpu          → 1
#   mem           → 1024
#   schedule_time → 2022-11-18 11:35:08
#   create_time   → 0000-01-01 00:00:00
#   start_time    → 0000-01-01 00:00:00
#   stop_time     → 0000-01-01 00:00:00
#   wall_time     → 1 hour
#   state         → :queuing
#   priority      → 20
#   dependency    → 2 jobs
#   task          → Task
#   stdout_file   → ""
#   stderr_file   → ""
#   _thread_id    → 0
```

> `dependency` argument in `Job(...)` controls when to start a job.
>
> It is a vector with element `STATE => job` or `STATE => job.id`.
>
> STATE is one of `DONE`, `FAILED`, `CANCELLED`, `QUEUING`, `RUNNING`, `PAST`.  
> The first 5 states are real job states.  
> `PAST` is the super set of `DONE`, `FAILED`, `CANCELLED`, which means the job will not run in the future.

Submit a job to queue:

```julia
submit!(command_job)
submit!(task_job)
submit!(job_with_args)
```

Cancel or interrupt a job:

```julia
cancel!(command_job)
```

Get the returned result:

```julia
result(job_with_args)
# "result"
```

### Queue

Show queue (all jobs):
```julia
queue(:all)      # or:
queue(all=true)  # or:
all_queue()
# 3-element Vector{Job}:
# ┌─────┬───────┬──────────────────┬─────────────────┬──────┬──────┬────────
# │ Row │ state │               id │            name │ user │ ncpu │  mem  ⋯
# ├─────┼───────┼──────────────────┼─────────────────┼──────┼──────┼────────
# │   1 │ :done │ 3603962559801309 │              "" │   "" │    1 │    0  ⋯
# │   2 │ :done │ 3603962563817452 │              "" │   "" │    1 │    0  ⋯
# │   3 │ :done │ 3603962603799412 │ "job with args" │ "me" │    1 │ 1024  ⋯
# └─────┴───────┴──────────────────┴─────────────────┴──────┴──────┴────────
#                                                         11 columns omitted
```

Show queue (running and queuing jobs only):

```julia
queue()
# 0-element Vector{Job}:
```

Show queue using a job state (`QUEUING`, `RUNNING`, `DONE`, `FAILED`, `CANCELLED`, or `PAST`):

```julia
queue(DONE)
# 3-element Vector{Job}:
# ┌─────┬───────┬──────────────────┬─────────────────┬──────┬──────┬────────
# │ Row │ state │               id │            name │ user │ ncpu │  mem  ⋯
# ├─────┼───────┼──────────────────┼─────────────────┼──────┼──────┼────────
# │   1 │ :done │ 3603962559801309 │              "" │   "" │    1 │    0  ⋯
# │   2 │ :done │ 3603962563817452 │              "" │   "" │    1 │    0  ⋯
# │   3 │ :done │ 3603962603799412 │ "job with args" │ "me" │    1 │ 1024  ⋯
# └─────┴───────┴──────────────────┴─────────────────┴──────┴──────┴────────
#                                                         11 columns omitted
```

`queue(...)` and `all_queue()` can also used to filter job name and user. Please find out more by typing `?queue` and `?all_queue` in REPL.

### Job Query

Get `Job` object by providing job ID.

```julia
job_query(3603962563817452)  # or:
queue(3603962563817452)
# Job:
#   id            → 3603962563817452
#   name          → ""
#   user          → ""
#   ncpu          → 1
#   mem           → 0
#   schedule_time → 0000-01-01 00:00:00
#   create_time   → 2022-11-18 11:35:30
#   start_time    → 2022-11-18 11:35:30
#   stop_time     → 2022-11-18 11:35:30
#   wall_time     → 1 week
#   state         → :done
#   priority      → 20
#   dependency    → []
#   task          → Task
#   stdout_file   → ""
#   stderr_file   → ""
#   _thread_id    → 0
```

### Wait for jobs

Wait for all jobs in `queue()` finished using `wait_queue(; show_progress::Bool)`.

If `show_progress = true`, a fancy progress meter will display.

```julia
wait_queue()
# no output

wait_queue(show_progress = true)
```
![progress meter](assets/progress_meter.png)

### Find optimized `ncpu` that a Job can use

Only available from JobSchedulers v0.7.8.

```julia
solve_optimized_ncpu(default::Int; 
    ncpu_range::UnitRange{Int64} = 1:total_cpu, 
    njob::Int = 1, 
    total_cpu::Int = JobSchedulers.SCHEDULER_MAX_CPU, 
    side_jobs_cpu::Int = 0)
```

Find the optimized number of CPU for a job.

  - `default`: default ncpu of the job.
  - `ncpu_range`: the possible ncpu range of the job.
  - `njob`: number of the same job.
  - `total_cpu`: the total CPU that can be used by JobSchedulers.
  - `side_jobs_cpu`: some small jobs that might be run when the job is running, so the job won't use up all of the resources and stop small tasks.

### Compatibility with Pipelines.jl

[Pipelines.jl]: https://cihga39871.github.io/Pipelines.jl/dev/	"Pipelines.jl: A lightweight Julia package for computational pipelines and workflows."

You can also create a `Job` by using `Program` types from Pipelines.jl:

```julia
Job(p::Program; program_kwargs..., run_kwargs..., job_kwargs...)
Job(p::Program, inputs; run_kwargs..., job_kwargs...)
Job(p::Program, inputs, outputs; run_kwargs..., job_kwargs...)
```

- `program_kwargs...` is input and output arguments defined in `p::Program`.
- `run_kwargs...` is keyword arguments of `run(::Program; ...)`
- `job_kwargs...` is keyword arguments of `Job(::Union{Base.AbstractCmd,Task}; ...)`

Details can be found by typing

```julia
julia> using Pipelines, JobSchedulers
julia> ?run
julia> ?Job
```

#### Example

```julia
using Pipelines, JobSchedulers

scheduler_start()

p = CmdProgram(
    inputs = ["IN1", "IN2"],
    outputs = "OUT",
    cmd = pipeline(`echo inputs are: IN1 and IN2` & `echo outputs are: OUT`)
)
# CmdProgram:
#   name             → Command Program
#   id_file          → 
#   info_before      → auto
#   info_after       → auto
#   cmd_dependencies → CmdDependency[]
#   arg_inputs       → IN1 :: Any (required)
#                      IN2 :: Any (required)
#   validate_inputs  → do_nothing
#   prerequisites    → do_nothing
#   cmd              → `echo inputs are: IN1 and IN2` & `echo outputs are: OUT`
#   infer_outputs    → do_nothing
#   arg_outputs      → OUT :: Any (required)
#   validate_outputs → do_nothing
#   wrap_up          → do_nothing
#   arg_forward      → Pair{String, Symbol}[]


### native Pipelines.jl method to run the program
run(p, IN1 = `in1`, IN2 = 2, OUT = "out", touch_run_id_file = false) # touch_run_id_file = false means do not create a file which indicates the job is done and avoids re-run.

# inputs are: in1 and in2
# outputs are: out
# (true, Dict("OUT" => "out"))

### run the program by submitting to JobSchedulers.jl
program_job = Job(p, IN1 = `in1`, IN2 = 2, OUT = "out", touch_run_id_file = false)
# Job:
#   id            → 3603980229784158
#   name          → "Command Program"
#   user          → ""
#   ncpu          → 1
#   mem           → 0
#   schedule_time → 0000-01-01 00:00:00
#   create_time   → 0000-01-01 00:00:00
#   start_time    → 0000-01-01 00:00:00
#   stop_time     → 0000-01-01 00:00:00
#   wall_time     → 1 week
#   state         → :queuing
#   priority      → 20
#   dependency    → []
#   task          → Task
#   stdout_file   → ""
#   stderr_file   → ""
#   _thread_id    → 0

submit!(program_job)
# inputs are: in1 and 2
# outputs are: out

# get the returned result
result(program_job)
# (true, Dict("OUT" => "out"))
```

### Backup

Set backup file:

```julia
set_scheduler_backup("/path/to/backup/file")
```
> JobSchedulers writes to the backup file at exit.
> If the file exists, scheduler settings and job queue will be recovered from it automatically.
> Recovered jobs are just for query, not run-able.

Stop backup and `delete_old` backup:

```julia
set_scheduler_backup(delete_old=true)
```

Backup immediately:

```julia
backup()
```


CHANGELOG.md
# Changelog

v0.7.12

- Compat: Pipelines v0.9, 0.10 (new), 1 (not published).

v0.7.11

- Update: Term to v2.
- Feat: Set a lower loop interval of nthreads > 2.
- Feat: Move scheduler_start() in __init__

v0.7.10

- Feature: Better progress bar for visualization.

v0.7.9

- Fix: `solve_optimized_ncpu()`: devision by 0 if njob == 0.

v0.7.8

- Feature: `solve_optimized_ncpu()`: Find the optimized number of CPU for a job.

v0.7.7

- Fix: `style_line()`: index error for special UTF characters.

v0.7.6

- Fix: if original stdout is a file, not contaminating stdout using `wait_queue(show_progress = true)`.

v0.7.5

- Change: remove extra blank lines after `wait_queue(show_progress = true)`.

- Fix a benign error (task switch error for `sleep()`).

v0.7.4

- Feature: Progress meter: `wait_queue(show_progress = true)`.

v0.7.3

- Compat: Pipelines v0.9: significant improvement on decision of re-run: considering file change.
- Fix: pretty print of Job and Vector{Job}.

v0.7.2

- Fix: unexpected output of `scheduler_status()` when SCHEDULER_TASK is not defined.

v0.7.1

- Compat: PrettyTables = "0.12 - 2" to satisfy DataFrames v1.3.5 which needs PrettyTables v1 but not v2.

v0.7.0

- Remove dependency DataFrames and change to PrettyTables. The loading time of DataFrames is high.

- Feature: now a Job is sticky to one thread (>1). JobSchedulers allocates and manuages it. The SCHEDULER_TASK is sticky to thread 1.

- Feature: `queue(...)` is rewritten.

- Feature: Better pretty print of Job and queue().

- Feature: New function: `wait_queue()` waits for all jobs in `queue()` become finished.

- Feature: New function: `set_scheduler()`

- Fix: `set_scheduler_max_cpu(percent::Float64)`: use default_ncpu() if error.

- Change: SCHEDULER_UPDATE_SECOND to 0.05 from 0.6

v0.6.12

- Feature: Enchance compatibility with Pipelines v0.8.5: Program has a new field called arg_forward that is used to forward user-defined inputs/outputs to specific keyword arguments of JobSchedulers.Job(::Program, ...), including name::String, user::String, ncpu::Int, mem::Int.

v0.6.11

- Fix: running `queue()` when updating queue: use lock within `DataFrames.DataFrame(job_queue::Vector{Job})`.

v0.6.10

- Update documents.

v0.6.9

- Support Pipelines.jl v0.8.

v0.6.8

- Feature: Replace `@Job` with `Job` to run `program` without creating `inputs::Dict` and `outputs::Dict`. Remove `@Job`.

v0.6.7

- Feature: Run `program` without creating `inputs::Dict` and `outputs::Dict`: `@Job program::Program key_value_args... Job_args...`. See also `@run` in Pipelines.jl.

v0.6.6

- Optimize: `job.dependency` now accepts `DONE => job`, `[DONE => job1.id; PAST => job2]`.

- Optimize: `is_dependency_ok(job::Job)::Bool` is rewritten: for loop when found a dep not ok, and delete previous ok deps. If dep is provided as Int, query Int for job and then replace Int with the job.

v0.6.5

- Fix: If an app is built, SCHEDULER_MAX_CPU and SCHEDULER_MAX_MEM will be fixed to the building computer: fix by re-defining `SCHEDULER_MAX_CPU` and `SCHEDULER_MAX_MEM` in `__init__()`.

- Debug: add debug outputs.

v0.6.4

- Fix: `scheduler_stop()` cannot stop because v0.6.1 update. Now `scheduler_stop` does not send ^C to `SCHEDULER_TASK`, but a new global variable `SCHEDULER_WHILE_LOOP::Bool` is added to control the while loop in `scheduler()`.

- Optimize: the package now can be precompiled: global Task cannot be precompiled, so we do not define `SCHEDULER_TASK::Task` when loading the package. Define it only when needed.

v0.6.3

- Fix: `scheduler_start()` now wait until `SCHEDULER_TASK` is actually started. Previously, it returns after `schedule(SCHEDULER_TASK)`.

v0.6.2

- Compat Pipelines v0.7.0.

v0.6.1

- Robustness: scheduler() and wait_for_lock(): wrap sleep() within a try-catch block. If someone sends ctrl + C to sleep, scheduler wont stop.

v0.6.0

- Compatibility: Pipelines v0.5.0: Job(...; dir=dir).

v0.5.1

- Fix: program_close_io: If the current stdout/stderr is IO, restore to default stdout/stderr.

v0.5.0

- Compatibility: Pipelines v0.5.0: fixed redirection error and optimized stack trace display. Extend `Base.istaskfailed` to fit Pipelines and JobSchedulers packages, which will return a `StackTraceVector` in `t.result`, while Base considered it as `:done`. The fix checks the situation and modifies the real task status and other properties.

v0.4.1

- Export `PAST`. PAST is the super set of DONE, FAILED, CANCELLED, which means the job will not run in the future.

v0.4.0

- If running with multi-threads Julia, `SCHEDULER_TASK` runs in thread 1, and other jobs spawn at other threads. Thread assignment was achieved by JobScheduler. Besides, `SCHEDULER_MAX_CPU = nthreads() > 1 ? nthreads()-1 : Sys.CPU_THREADS`.

- New feature: `queue(job_state::Symbol)`.

- Use try-finally for all locks.

v0.3.0

- Tasks run on different threads, if Julia version supports and `nthreads() > 1`.

- Use `SpinLock`.

- Fix typo "queuing" from "queueing".

- Notify when a job is failed.


src/control.jl

# TODO atexit(f) to store the job history.


function new_scheduler_task()
    global SCHEDULER_TASK
    SCHEDULER_TASK = @task scheduler()
    @static if :sticky in fieldnames(Task)
        # make the scheduler task sticky to threadid == 1
        if nthreads() > 1
            # sticky: disallow task migration which was introduced in 1.7
            @static if VERSION >= v"1.7"
                SCHEDULER_TASK.sticky = true
            else
                SCHEDULER_TASK.sticky = false
            end    
            ccall(:jl_set_task_tid, Cvoid, (Any, Cint), SCHEDULER_TASK, 0)
        end
    end
    SCHEDULER_TASK
end

"""
    scheduler_start()

Start the job scheduler.
"""
function scheduler_start(; verbose=true)
    global SCHEDULER_TASK

    if !isdefined(@__MODULE__, :SCHEDULER_TASK)
        new_scheduler_task()
    end

    set_scheduler_while_loop(true) # scheduler task won't stop

    if istaskfailed(SCHEDULER_TASK) || istaskdone(SCHEDULER_TASK)
        verbose && @warn "Scheduler was interrupted or done. Restart."
        new_scheduler_task()
        schedule(SCHEDULER_TASK)
        while !istaskstarted(SCHEDULER_TASK)
    		sleep(0.05)
    	end
    elseif istaskstarted(SCHEDULER_TASK) # if done, started is also true
        verbose && @info "Scheduler is running."
    else
        verbose && @info "Scheduler starts."
        schedule(SCHEDULER_TASK)
        while !istaskstarted(SCHEDULER_TASK)
    		sleep(0.05)
    	end
    end
end

"""
    scheduler_stop()

Stop the job scheduler.
"""
function scheduler_stop(; verbose=true)
    global SCHEDULER_TASK

    if !isdefined(@__MODULE__, :SCHEDULER_TASK)
        verbose && @warn "Scheduler is not running."
    elseif istaskfailed(SCHEDULER_TASK) || istaskdone(SCHEDULER_TASK)
        verbose && @warn "Scheduler is not running."
    elseif istaskstarted(SCHEDULER_TASK) # if done, started is also true
        set_scheduler_while_loop(false) # scheduler task stop after the next loop
        while !(istaskfailed(SCHEDULER_TASK) || istaskdone(SCHEDULER_TASK))
            sleep(0.2)
        end
        verbose && @info "Scheduler stops."
    else
        verbose && @warn "Scheduler is not running."
    end
end

"""
    scheduler_status() :: Symbol

Print the settings and status of job scheduler. Return `:not_running` or `:running`.
"""
function scheduler_status(; verbose=true)
    global SCHEDULER_TASK
    global SCHEDULER_MAX_CPU
    global SCHEDULER_MAX_MEM
    global SCHEDULER_UPDATE_SECOND
    global JOB_QUEUE_MAX_LENGTH
    global SCHEDULER_TASK
    if !isdefined(@__MODULE__, :SCHEDULER_TASK)
        verbose && @warn "Scheduler is not running." SCHEDULER_MAX_CPU SCHEDULER_MAX_MEM SCHEDULER_UPDATE_SECOND JOB_QUEUE_MAX_LENGTH
        :not_running
    elseif istaskfailed(SCHEDULER_TASK) || istaskdone(SCHEDULER_TASK)
        verbose && @info "Scheduler is not running." SCHEDULER_MAX_CPU SCHEDULER_MAX_MEM SCHEDULER_UPDATE_SECOND JOB_QUEUE_MAX_LENGTH SCHEDULER_TASK
        :not_running
    elseif istaskstarted(SCHEDULER_TASK)
        verbose && @info "Scheduler is running." SCHEDULER_MAX_CPU SCHEDULER_MAX_MEM SCHEDULER_UPDATE_SECOND JOB_QUEUE_MAX_LENGTH SCHEDULER_TASK
        :running
    else
        verbose && @info "Scheduler is not running." SCHEDULER_MAX_CPU SCHEDULER_MAX_MEM SCHEDULER_UPDATE_SECOND JOB_QUEUE_MAX_LENGTH SCHEDULER_TASK
        :not_running
    end
end

"""
    set_scheduler_update_second(s::Float64 = 0.6)

Set the update interval of scheduler.
"""
function set_scheduler_update_second(s::Float64 = 0.6)
    s <= 0.001 && error("schedular update interval cannot be less than 0.001.")
    global SCHEDULER_UPDATE_SECOND = s
end
set_scheduler_update_second(s) = set_scheduler_update_second(convert(Float64, s))


default_ncpu() = ifelse(nthreads() > 1, nthreads()-1, Sys.CPU_THREADS)

"""
    set_scheduler_max_cpu(ncpu::Int = default_ncpu())
    set_scheduler_max_cpu(percent::Float64)

Set the maximum CPU (thread) the scheduler can use. If starting Julia with multi-threads, the maximum CPU is `nthreads() - 1`.

# Example
    set_scheduler_max_cpu()     # use all available CPUs
    set_scheduler_max_cpu(4)    # use 4 CPUs
    set_scheduler_max_cpu(0.5)  # use 50% of CPUs
"""
function set_scheduler_max_cpu(ncpu::Int = default_ncpu())
    ncpu < 1 && error("number of CPU cannot be less than 1.")
    if ncpu > Sys.CPU_THREADS
        @warn "Assigning number of CPU > total CPU."
    end
    # nthreads == 1 will not triger schedule at different jobs.
    if nthreads() > 1
        if ncpu > nthreads()-1
            error("Assigning number of CPU > total threads available - 1. Thread one is reserved for schedulers. Try to start Julia with sufficient threads. Help: https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads")
        elseif ncpu < 2
            error("Assigning number of CPU < 2 is not allowed in multi-threaded Julia. Thread one is reserved for schedulers.")
        end
    end
    global SCHEDULER_MAX_CPU = ncpu
end
function set_scheduler_max_cpu(percent::Float64)
    if 0.0 < percent <= 1.0
        ncpu = round(Int, Sys.CPU_THREADS * percent)
        if ncpu > default_ncpu()
            @warn "Assigning number of CPU > default_ncpu() is not allowed. Set to default_ncpu(). Thread one is reserved for schedulers if Threads.nthreads() > 1. To use more threads, try to start Julia with sufficient threads. Help: https://docs.julialang.org/en/v1/manual/multi-threading/#Starting-Julia-with-multiple-threads"
            ncpu = default_ncpu()
        end
        set_scheduler_max_cpu(ncpu)
    else
        @error "Percent::Float64 should be between 0 and 1. Are you looking for set_scheduler_max_cpu(ncpu::Int) ?"
    end
end

default_mem() = round(Int, Sys.total_memory() * 0.8)
"""
    set_scheduler_max_mem(mem::Int = default_mem())
    set_scheduler_max_mem(percent::Float64)

Set the maximum RAM the scheduler can use.

# Example
    set_scheduler_max_mem()             # use 80% of total memory

    set_scheduler_max_mem(4GB)          # use 4GB memory
    set_scheduler_max_mem(4096MB)
    set_scheduler_max_mem(4194304KB)
    set_scheduler_max_mem(4294967296B)

    set_scheduler_max_mem(0.5)          # use 50% of total memory
"""
function set_scheduler_max_mem(mem::Int = default_mem())
    mem < 1 && error("number of memory cannot be less than 1.")
    if mem > Sys.total_memory() * 0.9
        @warn "Assigning memory > 90% of total memory."
    end
    global SCHEDULER_MAX_MEM = mem
end
function set_scheduler_max_mem(percent::Float64)
    if 0.0 < percent < 1.0
        set_scheduler_max_mem(round(Int, Sys.total_memory() * percent))
    else
        @error "Percent::Float64 should be between 0 and 1. Are you looking for set_scheduler_max_mem(mem::Int) ?"
    end
end

"""
    set_scheduler_max_job(n_finished_jobs::Int = 10000)

Set the number of finished jobs.
"""
function set_scheduler_max_job(n_finished_jobs::Int = 10000)
    if n_finished_jobs < 10
        @error "Cannot set number of finished jobs < 10"
    else
        global JOB_QUEUE_MAX_LENGTH = n_finished_jobs
    end
end

"""
    set_scheduler(;
        max_cpu::Union{Int,Float64} = JobSchedulers.SCHEDULER_MAX_CPU,
        max_mem::Union{Int,Float64} = JobSchedulers.SCHEDULER_MAX_MEM,
        max_job::Int = JobSchedulers.JOB_QUEUE_MAX_LENGTH,
        update_second = JobSchedulers.SCHEDULER_UPDATE_SECOND
    )

See details:
[`set_scheduler_max_cpu`](@ref), 
[`set_scheduler_max_mem`](@ref), 
[`set_scheduler_max_job`](@ref), 
[`set_scheduler_update_second`](@ref) 
"""
function set_scheduler(;
    max_cpu::Union{Int,Float64} = SCHEDULER_MAX_CPU,
    max_mem::Union{Int,Float64} = SCHEDULER_MAX_MEM,
    max_job::Int = JOB_QUEUE_MAX_LENGTH,
    update_second = SCHEDULER_UPDATE_SECOND
)
    set_scheduler_max_cpu(max_cpu)
    set_scheduler_max_mem(max_mem)
    set_scheduler_max_job(max_job)
    set_scheduler_update_second(update_second)

    scheduler_status()
end

"""
    wait_queue(;show_progress = false)

Wait for all jobs in `queue()` become finished.

- `show_progress = true`, job progress will show.

See also: [`queue_progress`](@ref).
"""
function wait_queue(;show_progress::Bool = false)
    if show_progress
        queue_progress()
    else
        while length(JOB_QUEUE) > 0 && scheduler_status(verbose=false) === RUNNING
            sleep(SCHEDULER_UPDATE_SECOND)
        end
        if scheduler_status(verbose=false) != RUNNING
            @error "Scheduler was not running. Jump out from wait_queue()"
        end
    end
    nothing
end